src/Scientist/Internals/ExperimentResultComparer.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GitHub.Internals
+{
+    internal class ExperimentResultComparer<T> : IEqualityComparer<ExperimentInstance<T>.ExperimentResult>
+    {
+        private readonly Func<T, T, bool> _resultComparison;
+        private readonly IEqualityComparer<T> _resultEqualityComparer;
+
+        public ExperimentResultComparer(IEqualityComparer<T> resultEqualityComparer, Func<T, T, bool> resultComparison)
+        {
+            _resultEqualityComparer = resultEqualityComparer;
+            _resultComparison = resultComparison;
+        }
+
+
+        /// <summary>
+        /// Checks if two ExperimentResults are equal
+        /// </summary>
+        /// <param name="controlResult">Control ExperimentResult</param>
+        /// <param name="candidateResult">Candidate ExperimentResult</param>
+        /// <returns>
+        ///  Returns true if: 
+        /// 
+        /// 
+        ///  The values of the observations are equal according to an IEqualityComparer&lt;T&gt; expression, if given 
+        ///  The values of the observations are equal according to a comparison function, if given  
+        ///  Both observations raised an exception with the same Type and message.
+        ///  The values of the observation are both null 
+        ///  The values of the observations are equal according to Ts IEquatable&lt;T&gt; implementation, if implemented 
+        ///  The values of the observations are equal (using .Equals()) 
+        ///  
+        ///  Returns false otherwise. 
+        /// </returns>
+        public bool Equals(ExperimentInstance<T>.ExperimentResult controlResult, ExperimentInstance<T>.ExperimentResult candidateResult)
+        {
+            if (_resultComparison != null)
```

Comment:
I would like to refactor this a bit more but this is good enough for now.
I don't like having a whole bunch of `if(){return xyz()}`
(I would refactor into some sort of strategy pattern where all tests have to not fail or something)